File: 10-heap/00703-Kth_Largest_Element_In_A_Stream/00703-kth_largest_element_in_a_stream.cpp
// 703: Kth Largest Element In A Stream
// https://leetcode.com/problems/kth-largest-element-in-a-stream


#include <iostream>
#include <queue>
#include <vector>

using namespace std;


class KthLargest {
// SOLUTION
private:
    priority_queue<int, vector<int>, greater<int>> pq;
    int size;

public:
    KthLargest(int k, vector<int>& nums) {
        this->size = k;
        for (auto n : nums) {
            pq.push(n);
            if (pq.size() > k) pq.pop();
        }
    }

    int add(int val) {
        pq.push(val);
        if (pq.size() > size) pq.pop();
        return pq.top();
    }
};


int main() {
    // INPUT
    int k = 3;
    vector<int> nums = {4,5,8,2};
    KthLargest o(k, nums);


    // OPERATIONS
    cout << o.add(3) << endl;
    cout << o.add(5) << endl;
    cout << o.add(10) << endl;
    cout << o.add(9) << endl;
    cout << o.add(4) << endl;

    return 0;
}

File: 10-heap/00703-Kth_Largest_Element_In_A_Stream/00703-kth_largest_element_in_a_stream.java
// 703: Kth Largest Element In A Stream
// https://leetcode.com/problems/kth-largest-element-in-a-stream


import java.util.PriorityQueue;

class KthLargest {
    // SOLUTION
    static PriorityQueue<Integer> pq;
    static int size;

    public KthLargest(int k, int[] nums) {
        size = k;
        pq = new PriorityQueue<>(k);
        for (int n : nums)
            add(n);
    }

    public int add(int val) {
        if (pq.size() < size)
            pq.offer(val);
        else if (pq.peek() < val) {
            pq.poll();
            pq.offer(val);
        }
        return pq.peek();
    }

    
    public static void main(String[] args) {
        // INPUT
        int k = 3;
        int[] nums = {4,5,8,2};
        KthLargest o = new KthLargest(k, nums);

        // OPERATIONS
        System.out.println(o.add(3));
        System.out.println(o.add(5));
        System.out.println(o.add(10));
        System.out.println(o.add(9));
        System.out.println(o.add(4));
    }
}
